File: lintcode/153.combination-sum-ii.cpp
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
  /**
   * @param num: Given the candidate numbers
   * @param target: Given the target number
   * @return: All the combinations that sum to target
   */
  vector<vector<int>> combinationSum2(vector<int> &num, int target) {
    vector<vector<int>> res;
    vector<int> out;
    sort(num.begin(), num.end());
    combinationSum2DFS(num, res, out, 0, target);
    return res;
  }

  void combinationSum2DFS(const vector<int> &num, vector<vector<int>> &res,
                          vector<int> &out, int start, int target) {
    if (target == 0) {
      res.push_back(out);
      return;
    } else if (target < 0) {
      return;
    } else {
      for (int i = start; i < num.size(); ++i) {
        if (i > start && num[i] == num[i - 1]) {
          continue;
        }
        out.push_back(num[i]);
        combinationSum2DFS(num, res, out, i + 1, target - num[i]);
        out.pop_back();
      }
    }
  }
};

File: lintcode/153.combination-sum-ii.py
class Solution:
    """
    @param num: Given the candidate numbers
    @param target: Given the target number
    @return: All the combinations that sum to target
    """

    def combinationSum2(self, num, target):
        output = []
        res = []

        def combinationSum2DFS(num, target, start):
            if target == 0:
                res.append(list(output))
                return
            elif target < 0:
                return
            else:
                for i in range(start, len(num)):
                    if i > start and num[i] == num[i - 1]:
                        continue
                    output.append(num[i])
                    combinationSum2DFS(num, target - num[i], i + 1)
                    output.pop()

        num.sort()
        combinationSum2DFS(num, target, 0)
        return res


if __name__ == '__main__':
    s = Solution()
    print(s.combinationSum2([7, 1, 2, 5, 1, 6, 10], 8))
